app/services/work_packages/set_attributes_service.rb

```
@@ -51,17 +51,85 @@ def set_static_attributes(attributes)
   end
 
   def set_calculated_attributes(attributes)
+    derivable = derivable_attribute
     if work_package.new_record?
       set_default_attributes(attributes)
     else
       update_dates
     end
     update_duration
+    update_derivable(derivable)
     update_project_dependent_attributes
     reassign_invalid_status_if_type_changed
     set_templated_description
   end
 
+  def derivable_attribute
+    derivable_attribute_by_others_presence || derivable_attribute_by_others_absence
+  end
+
+  # Returns a field derivable by the presence of the two others.
+  #
+  # A field is derivable if it has not been set explicitly while the other two
+  # fields are set. Matching is done in the order :duration, :due_date,
+  # :start_date. The first one to match is returned.
+  def derivable_attribute_by_others_presence
```

What if all 3 parameters are set at once, and also incorrectly? Ie:
```ruby
start_date: Date.today, due_date: Date.tomorrow, duration: 10
```

It seems to flow through the `update_derivable` method without making a change.

---

Yes, my take on it is that the `UpdateContract` should raise an error but I'm not quite sure about it.

---

I added a checklist item on the work package so we do not forget it.